### PR TITLE
More stable and faster implementation of the gradient of torch.linalg.eigh 

### PR DESCRIPTION
**Pull request by lezcano**
This PR:
- Renames symeig_backward to eigh_backward
- Improves the stability and speed of the gradient computation by doing `V(A + B)Vh` instead of `VAVh + VBVh`  when both the gradients of the eigenvectors and eigenvalues are defined.
- Updates the comments of the function to make them arguably clearer

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
 
   auto glambda = grads[0];
   auto gv = grads[1];
 
   auto vh = v.conj().transpose(-2, -1);
 
+  if (glambda.defined()) {
+    // glambda is real even when the matrix is complex
+    glambda = glambda.to(self.dtype());
+  }
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2306]: It could be that there is no need to convert `glambda` to complex at all.
Elementwise products of complex-real tensors work and also `copy` seems to work as well:
```python
In [2]: import torch

In [3]: x = torch.rand(3, 3, dtype=torch.cdouble)

In [4]: y = torch.rand(3, 3, dtype=torch.double)

In [5]: x
Out[5]: 
tensor([[0.5320+0.1139j, 0.9173+0.0720j, 0.8534+0.6649j],
        [0.0010+0.0615j, 0.2283+0.1710j, 0.8142+0.7606j],
        [0.3496+0.3054j, 0.3624+0.1556j, 0.0159+0.5936j]],
       dtype=torch.complex128)

In [6]: x.copy_(y)
Out[6]: 
tensor([[0.9531+0.j, 0.6419+0.j, 0.2485+0.j],
        [0.2854+0.j, 0.4344+0.j, 0.2342+0.j],
        [0.0689+0.j, 0.7904+0.j, 0.3975+0.j]], dtype=torch.complex128)

In [7]: x
Out[7]: 
tensor([[0.9531+0.j, 0.6419+0.j, 0.2485+0.j],
        [0.2854+0.j, 0.4344+0.j, 0.2342+0.j],
        [0.0689+0.j, 0.7904+0.j, 0.3975+0.j]], dtype=torch.complex128)

```

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
-    glambda = glambda.to(self.dtype());
-    // computes v @ diag(glambda) @ vh
-    Tensor glambda_term = at::matmul(v * glambda.unsqueeze(-2), vh);
-    if (at::inplaceIsVmapCompatible(result, glambda_term)) {
-      result.add_(glambda_term);
+    result = at::matmul(v, at::matmul(result, vh));
+  }
+  else {
+    if (glambda.defined()) {
+      // If we just glambda is defined, one matmul suffices
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2332]: `If we just glambda is defined` -> `If only glambda is defined`.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+    if (at::GradMode::is_enabled()) {
+      // Avoids differentiating through at infinity when doing gradgrad
+      // 1 could be any number, as we are going to overwrite the diagonal
+      F.diagonal(/*offset=*/0, /*dim1=*/-2, /*dim2=*/-1).fill_(1);
+    }
+    // We compute first the multiplication F*(vh @ gv) and then we zero-out the diagonal.
+    // Otherwise, if vh @ gh has NaNs on the diagonal, these would be propagated to the result
+    // as NaN * 0 = NaN.
+    result =  at::matmul(vh, gv) / F;
+    if (glambda.defined()) {
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2321]: I disagree that the gradient is symmetric, primarily because the matrix `F` is NOT symmetric.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+      result = at::matmul(v * glambda.unsqueeze(-2), vh);
     } else {
-      result = result + glambda_term;
+      // If neither is defined, there's nothing to do
+      return at::zeros_like(self, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
     }
   }
+  // The result here is alreaday symmetric (resp. Hermitian), but we project it onto the
+  // symmetric (resp. Hermitian) matrices to account for numerical errors
   return result.add(result.conj().transpose(-2, -1)).mul_(0.5);
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2341]: I disagree that the gradient is symmetric, primarily because the matrix `F` is NOT symmetric. But we assume that the elements of the input matrix, for real symmetric case, for example, are equivalent, i.e. `a_ij == a_ji`, so, even if the resulting gradient result is NOT symmetric, we can symmetrize it.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2321]: F is skew, and `vh @ gv` is skew (or skew-Hermitian), so the result is symmetric. Looking at the derivation you can also prove that its diagonal is zero, which is why we can do the copy below.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 5]
```diff
@@ -2276,50 +2278,70 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
-  // were updated by a gradient based algorithm.
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
+Tensor eigh_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
+                     bool eigenvectors, const Tensor& lambda, const Tensor& v) {
+  // This function is used for both torch.symeig and torch.linalg.eigh.
+  // eigh (and torch.symeig) operates only on symmetric (resp. Hermitian) inputs.
+  // This is a _linear_ subspace of R^{n x n} (resp. C^{n x n}),
+  // since A, B \in Sym(n) => A + B \in Sym(n)
+  // This means that, the tangent space to Sym(n) is Sym(n),
+  // so the gradient is symmetric (resp. Hermitian).
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2289]: Is it really true? Note that the matrix `F` below is NOT symmetric, and I am not sure that the numerical argument for the gradient symmetrization holds...

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 3 above)
Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:2321]: Hmm, why is that `vh @ gv` is skew-symmetric?

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2321]: If that is the case, what we should do is project `vh @ gv` to be skew-symmetric. In theory, `vh @ gv` should be skew-symmetric since if you differentiate `V^T V = Id` for `V \in R^{n x k}` with orthonormal columns and `n >= k`, you get that `V^T  gv + gv^T V = 0`, that is, `V^T gv` is skew-symmetric.

Comment (nikitaved) [reply torch/csrc/autograd/FunctionsManual.cpp:2321]: As discussed, yes, `V^T dV` is indeed skew-symmetric, but `dV` and `gv` are completely different beasts. `dV` comes from the forward pass (differential), and `gv` comes from the backward pass (it is a gradient).

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2306]: Good point!

**Comment by Lezcano** [comment]
After some thinking, I understand and explained a bit better what's going on. In particular, the new implementation does a better job at explaining why can one copy the gradient wrt. lambda into the diagonal, without needing a `sum`.

As discussed offline, it was not very clear why the projection onto the symmetric matrices was correct. This implementation puts forward some ideas that should help explaining what's going on. In particular, it gets rid of this odd thing.

@nikitaved tell me what you think. I think that I could follow a very similar pattern on top of this implementation to implement the backwards of `torch.eig`. I'll do that once we got this one sorted.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 6]
```diff
@@ -2333,49 +2335,91 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+  // This function is used for both torch.symeig and torch.linalg.eigh.
+  // eigh (and torch.symeig) operates only on symmetric (resp. Hermitian) inputs.
+
+  // General considerations of the differential and adjoint
+  // Let U(n) = {U \in C^{n x n} | U^H U = I} by the unitary group and
+  // Her(n) = {A \in C^{n x n} | A^H = A} be the Hermitian matrices
+  // eigh : Her(n) -> U(n) x R^n
+  // Denoting the tangent spaces as T, the differential of eigh at A = VLV^H
+  // (i.e. forward differentiation) is a linear map
+  // (d eigh)_A : T_A Her(n) -> T_V U(n) x T_L R^n
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2349]: I am confused with the notation. Could you please clarify this line more?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 7]
```diff
@@ -2498,7 +2542,7 @@ Tensor linalg_qr_backward(const std::vector<torch::autograd::Variable> &grads, c
 }
 
 // Invertible case is derived from Jacobi's formula, and also can be found at:
-// http://eprints.maths.ox.ac.uk/1079/1/NA-08-01.pdf
+// https://people.maths.ox.ac.uk/gilesm/files/NA-08-01.pdf
```
Comment (nikitaved) [inline torch/csrc/autograd/FunctionsManual.cpp:2545]: A leak from a different branch? :)

Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2545]: I corrected the old URL (which does not work) to one that works.

**Review by nikitaved — APPROVED** [review]
LGTM! Thank you, Mario! Just a single minor nit regarding leaking of one branch (det) into this one.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 7 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2545]: It's just these two. That being said, I'm going to revert this one, as we are upgrading the derivative for the determinant soon anyway. We can change it then really.

**Comment by Lezcano** [comment]
Reverted the det change. This should be ready for review.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 8]
```diff
@@ -2086,9 +2087,10 @@ Tensor slice_backward_wrapper(
 // This makes no assumption on the signs of sigma.
 Tensor svd_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
           bool some, bool compute_uv, const Tensor& raw_u, const Tensor& sigma, const Tensor& raw_v) {
+
   TORCH_CHECK(compute_uv,
-           "svd_backward: Setting compute_uv to false in torch.svd doesn't compute singular matrices, ",
-           "and hence we cannot compute backward. Please use torch.svd(compute_uv=True)");
+           "svd_backward: Setting compute_uv=False in torch.svd doesn't compute the singular vectors, ",
```
Comment (mruberry) [inline torch/csrc/autograd/FunctionsManual.cpp:2092]: Language nit:

"cannot compute the backward" -> "cannot compute the gradient"

also, shouldn't this now direct people to torch.linalg.svdvals?

File: test/test_autograd.py [hunk 9]
```diff
@@ -3735,14 +3735,14 @@ def test_eig_complex_eigenvalues(self):
...
-        with self.assertRaisesRegex(RuntimeError, 'cannot compute backward'):
+        with self.assertRaisesRegex(RuntimeError, 'cannot compute the backward'):
             torch.autograd.backward([w, v], [torch.ones_like(w), torch.ones_like(v)])
 
     @skipIfNoLapack
     def test_svd_no_singularvectors(self):
         A = torch.randn(2, 2, dtype=torch.float32, requires_grad=True)
         u, s, v = torch.svd(A, compute_uv=False)
-        with self.assertRaisesRegex(RuntimeError, 'cannot compute backward'):
+        with self.assertRaisesRegex(RuntimeError, 'cannot compute the backward'):
```
Comment (mruberry) [inline test/test_autograd.py:3745]: "the backward" -> "the gradient"

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 10]
```diff
@@ -2326,48 +2328,91 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
+  //   - E_ij = L_i - L_j if i != j
+  //   - diag_embed takes a vector into a diagonal matrix
+  //   - The division by E is done just outside of the diagonal. In the diagonal it is set to zero
+
+  // This check just can be triggered in the backwards of torch.symeig
   TORCH_CHECK(eigenvectors,
-           "symeig_backward: Setting eigenvectors to false in torch.symeig doesn't compute eigenvectors ",
-           "and hence we cannot compute backward. Please use torch.symeig(eigenvectors=True)");
+           "Setting eigenvectors=False in torch.symeig doesn't compute the eigenvectors, ",
+           "and hence we cannot compute the backward. Please use torch.symeig(eigenvectors=True)");
```
Comment (mruberry) [inline torch/csrc/autograd/FunctionsManual.cpp:2374]: This note also needs an update since we just deprecated torch.symeig

**Comment by mruberry** [comment]
I suggested a couple docs updates inline to fit with our latest language and deprecations; let's see if @albanD's interested in sanity checking the autograd formula, too

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 8 above)
Comment (Lezcano) [reply torch/csrc/autograd/FunctionsManual.cpp:2092]: Sorry, this change does not belong in this PR. I reverted it.

File: test/test_autograd.py (hunk 9 above)
Comment (Lezcano) [reply test/test_autograd.py:3745]: Reverthing these as these changes are also done in the `symeig` deprecation PR.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 11]
```diff
@@ -2326,48 +2328,91 @@ Tensor eig_backward(const std::vector<torch::autograd::Variable> &grads, const T
...
-  // This gradient is symmetric, and not triangular.
-  // symeig operates only on symmetric inputs, which is a subspace of
-  // R^{n x n}, and hence the derivative is not well-defined for off-diagonal
-  // elements. We resolve this by taking the gradient of the functionally independent
-  // elements of the matrix (i.e., the lower triangular portion of the input) and then
-  // reflect it on the upper triangular portion, thereby symmetrizing the gradient of
-  // the symeig operation. The motivation behind this choice is that symmetric gradient
-  // leads to stable gradient updates, and retains symmetry of the updated matrix if it
-  // were updated by a gradient based algorithm.
+Tensor eigh_backward(const std::vector<torch::autograd::Variable> &grads, const Tensor& self,
```
[the remark this answers is not in the document]
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:2330]: I'm afraid I won't have time to dive into that much math right now :/
But if the tests are good, I trust gradcheck to be doing the right thing!

**Review by mruberry — APPROVED** [review]

**Comment by mruberry** [comment]
Unfortunately this is going to need a rebase; no rush as today is branch cut day (for better or worse), so landing will be extremely tricky for a day or two

**Comment by Lezcano** [comment]
@mruberry rebased

**Comment by mruberry** [comment]
@Lezcano looks like a lot of failures?

**Comment by Lezcano** [comment]
@mruberry sorry for that, now it should be fine
